pennylane/labs/lie/lie_closure.py

```diff
@@ -0,0 +1,133 @@
+            B -= np.trace(np.dot(C.conj().T, A)) * C
+        if np.linalg.norm(B) > tol:  # Tolerance for numerical stability
+            B /= np.linalg.norm(B)
+            basis.append(B)
+    return np.array(basis)
+
+
+def lie_closure_dense(
+    generators: Iterable[Union[PauliWord, PauliSentence, Operator]],
+    n=None,
+    max_iterations: int = 10000,
+    verbose: bool = False,
+    tol: float = None,
+):
+    r"""Compute the dynamical Lie algebra :math:`frak{g}` from a set of generators using their dense matrix representation.
+
+    This function computes the Lie closure of a set of generators using their dense matrix representation.
+    This is sometimes more efficient than using the sparse Pauli representations of :class:`~PauliWord` and
+    `~PauliSentence` that are employed in :func:`~lie_closure`, e.g., when ther are few but dense sums of Paulis.
+
+    .. seealso:: For details on the mathematical definitions, see :func:`~lie_closure` and our `Introduction to Dynamical Lie Algebras for quantum practitioners <https://pennylane.ai/qml/demos/tutorial_liealgebra/>`__.
+
+    Args:
+        generators (Iterable[Union[PauliWord, PauliSentence, Operator]]): generating set for which to compute the
+            Lie closure.
+        n (int): The number of qubits involved. If ``None`` is provided, it is automatically deduced from the generators.
+        max_iterations (int): maximum depth of nested commutators to consider. Default is ``10000``.
+        verbose (bool): whether to print out progress updates during Lie closure
+            calculation. Default is ``False``.
+        tol (float): Numerical tolerance for the linear independence check used in :class:`~.PauliVSpace`.
+
+    Returns:
+        numpy.ndarray: The ``(dim(g), 2**n, 2**n)`` array containing the linear independent basis of the DLA g as dense matrices.
+
+    **Example**
+
+    Compute the Lie closure of the isotropic Heisenberg model with generators :math:`\{X_i X_{i+1} + Y_i Y_{i+1} + Z_i Z_{i+1}}_{i=0}^{n-1}`.
+
+    >>> n = 5
+    >>> gens = [X(i) @ X(i+1) + Y(i) @ Y(i+1) + Z(i) @ Z(i+1) for i in range(n-1)]
+    >>> g = lie_closure_mat(gens, n)
+
+    The result is a ``numpy`` array. We can turn the matrices back into PennyLane operators by employing :func:`~pauli_decompose`.
+
+    >>> g_ops = [qml.pauli_decompose(op) for op in g]
+
+    """
+
+    if n is None:
+        n = len(qml.wires.Wires.all_wires([_.wires for _ in generators]))
+
+    gens = np.array([qml.matrix(op, wire_order=range(n)) for op in generators], dtype=complex)
+    _, chi, _ = gens.shape
+    vspace = _hermitian_basis(gens, tol)
+
+    epoch = 0
+    old_length = 0  # dummy value
+    new_length = len(vspace)
+
+    while (new_length > old_length) and (epoch < max_iterations):
+        if verbose:
+            print(f"epoch {epoch+1} of lie_closure_dense, DLA size is {new_length}")
+
+        # compute all commutators
+        m0m1 = np.einsum("aij,bjk->abik", vspace, gens)
```

Comment:
Also, we should explore whether we want `tensordot` or `einsum`, I suppose :)

Reply:
I got defeated by tensordot and eventually caved to einsum 😢 

> I also was thinking: We're re-computing commutators if we take the full guys here, right?
The following skips over the pairs of commutators that we already computed earlier:

Oh yeah there definitely is redundancy, good point! In the old implementation we had `combinations(vspace, 2)`.
I'm still tempted to do `vspace[old_length:], gens` because that "naturally" corresponds to the nesting depth, which enables a user to just do the closure until a certain level of nesting through `max_iterations`. What do you think? There is no redundancy but it _could_ be made faster by including more combinations at once.

Reply:
There is also redundancy in that we are computing both `A@B` and `B@A` twice, but I couldnt figure out how to re-order stuff to only have it once

Reply:
> I'm still tempted to do vspace[old_length:], gens because that "naturally" corresponds to the nesting depth, which enables a user to just do the closure until a certain level of nesting through max_iterations. What do you think? There is no redundancy but it could be made faster by including more combinations at once.

Ohhh, that is so neat, I had not thought of this. 
Let's go for this and document that this might be a speedup left on the road for this feature.

Reply:
>There is also redundancy in that we are computing both [A,B] and [B,A], but I couldnt figure out how to re-order stuff to only have it once. I figured thats still outweighed by the ability to vectorize the commutators

If we use the `[old_length:]` slicing I do not think that we have redundancy anymore :)

Reply:
> Let's go for this and document that this might be a speedup left on the road for this feature.

@Qottmann Actually, could we do something like `vspace[old_length:], vspace[:initial_length]`? Then we keep the minimal basis for the initial set of operators but also the "nesting level <> iteration" correspondence

Reply:
It's the very first length.
Your suggestion was to use `gens`, which have not run through `_hermitian_basis` yet, right?
So my suggestion is to use `_hermitian_basis(gens)`, which is coincidentally stored in `vspace[:initial_length]` with `initial_length=len(vspace)` _before_ the `while` loop

Reply:
ah gotcha, yes! makes sense